docs/proposals/provider-agnostic-dns-health-checks.md

```diff
@@ -0,0 +1,86 @@
+# Provider agnostic DNS Health checks
+
+## Introduction
+
+The MGC has the ability of extending the DNS configuration of the gateway with the DNSPolicy resource. This resource allows users to configure health checks. As a result of configuring health checks, the controller creates the health checks in Route53, attaching them to the related DNS records. This has the benefit of automatically disabling an endpoint if it becomes unhealthy, and enabling it again when it becomes healthy again.
+This feature has a few shortfalls:
+
+1. It’s tightly coupled with Route53. If other DNS providers are supported they must either provide a similar feature, or health checks will not be supported
+2. Lacks the ability to reach endpoints in private networks
+
+This document describes a proposal to extend the current health check implementation to overcome these shortfalls.
+
+### Goals
+
+* Ability to configure health checks in the DNSPolicy associated to a Gateway
+* Ability for the health checks to reach endpoints in private networks
+* DNS records are disabled when the associated health check fails
+* Current status of the defined health checks is visible to the end user
+* Transparently keep support for other health check providers like Route53
+
+## Proposal
+
+### Control plane
+
+Add a field in the health check section of the DNSPolicy to select the health
+check type to use
+
+```yaml
+healthCheck:
+  # ...
+  port: "..."
+  # ...
+  type: "Provider | Agent"
+```
+
+The presence of the `healthCheck` field in the DNSPolicy will affect the reconciliation
+of the endpoints when creating/updating DNSRecords:
+
+* For every endpoint that is generated, a health check is created based on the configuration in the DNSPolicy
+* Relevant `providerSpecific` fields are added to the endpoint depending on the
+  `type`:
+
+  * `Provider`: In the case of AWS, the `aws/health-check-id` field will be set with the ID of the health check in Route53
+  * `Agent`:
+    * A reference to the equivalent `DNSHealthCheckProbe` in the downstream cluster
+    will be set: `agent/probe-name`, `agent/probe-namespace`
+    * The health status of the `DNSHealthCheckProbe` will be propagated: `agent/health-status`
+
+
+### Data plane
+
+#### `DNSHealthCheckProbe` resource
+
+The DNSHealthCheckProbe resource configures the agent in the data plane cluster to perform the health checks against a local endpoint.
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: DNSHealthCheckProbe
+metadata:
+  name: example-probe
+spec:
+  port: "..."
+  host: “...”
+  ipAddress: "..."
```

Comment:
could this possibly just be set to Address?

Reply:
> and have one property for host and address?

Do you mean that the requests are made to `http(s)://<address>:<port><path>` instead of `http(s)://<ipAddress>:<port><path>` with the `Host: <host>` header